New container cost allocation documentation (#17514)
DataDog/documentation: content/en/cloud_cost_management/container_cost_allocation.md

---
title: Container Cost Allocation
kind: documentation
disable_toc: false
private: true
further_reading:
- link: "https://docs.datadoghq.com/cloud_cost_management/?tab=aws#overview"
  tag: "Documentation"
  text: "Cloud Cost Management"
---

{{< jqmath-vanilla >}}

## Overview

Datadog Cloud Cost Management (CCM) automatically allocates EC2 compute cost in Kubernetes and ECS clusters to individual pods and tasks running in those clusters. Use cost metrics enriched with tags from pods, nodes, containers, and tasks to visualize container workload cost in the context of your entire cloud bill.

{{< img src="cloud_cost/container_cost_allocation/cost_allocation_table.png" alt="Your image description" style="width:100%;" >}}

## Prerequisites

1. Set up and configure [AWS Cloud Cost integration][1].
1. At least one of the following must be running:
    - [**Datadog Kubernetes agent**][2] in a Kubernetes environment using EC2 instances. 

>     - **Datadog container monitoring** in containers running in ECS tasks. The container agent must be configured to use [orchestrator tag cardinality][3].

Reviewer: ```suggestion
    - **Datadog Container Monitoring** in containers running in ECS tasks. The container agent must be configured to use [orchestrator tag cardinality][3].
```

>     - **Datadog Container Monitoring** in containers running in ECS tasks. The container agent must be configured to use [orchestrator tag cardinality][3].

Reviewer 2: orchestrator tag cardinality may not actually be necessary. @jeannei did you get confirmation on this?

Reviewer 2: got confirmation from the #processes team that we can remove this req

## Cost metrics

When the prerequisites are met, new AWS cost metrics automatically appear.

| AWS Cost Metric                    | Description    |
| ---                                | ----------- | 
| `aws.cost.amortized.mem.allocated`   | EC2 costs allocated by memory requested by a pod or ECS task. <br> *Based on `aws.cost.amortized`* | 
| `aws.cost.net.amortized.mem.allocated` | Net EC2 costs allocated by memory requested by a pod or ECS task <br> *Based on `aws.cost.net.amortized`, if available* |
| `aws.cost.amortized.cpu.allocated` | EC2 costs allocated by CPU requested by a pod or ECS task <br> *Based on `aws.cost.amortized`* |

> | `aws.cost.net.amortized.cpu.allocated` | Net EC2 costs allocated by CPU requested by a pod or ECS task <br> *Based on `aws.cost.net.amortized`* |

Reviewer 3: Should we add `if available*` here as well since we have it for `aws.cost.net.amortized.mem.allocated`?

Author: ```suggestion
| `aws.cost.net.amortized.cpu.allocated` | Net EC2 costs allocated by CPU requested by a pod or ECS task <br> *Based on `aws.cost.net.amortized`, if available* |
```

These new cost metrics include all of your AWS cloud costs. This allows you to continue visualizing all of your cloud costs at one time, with added visibility into pods and tasks running on EC2 instances.

For example, say you have the tag `team` on S3 buckets, RDS stores, and Kubernetes pods. You can use one of the new metrics to group cost by `team`, and each group then includes the S3 and RDS costs for that team, as well as the cost of compute resources reserved by the tagged pods.

## Cost allocation

Cost allocation divides EC2 compute costs in the [Cost and Usage Report][4] (CUR) into individual costs for each pod or task running on the instance. These divided costs are then enriched with tags from the nodes, pods, and tasks, which allows you to break down costs by any associated dimensions.

### Kubernetes

> For the Kubernetes allocation, a Kubernetes node is joined with its associated EC2 instance costs. The node's cluster name and all node tags are added to the entire EC2 compute cost for the node. This allows you to associate cluster-level dimensions with the cost of the instance, without considering the pods scheduled to the node.

Reviewer: ```suggestion
For Kubernetes allocation, a Kubernetes node is joined with its associated EC2 instance costs. The node's cluster name and all node tags are added to the entire EC2 compute cost for the node. This allows you to associate cluster-level dimensions with the cost of the instance, without considering the pods scheduled to the node.
```

Next, Datadog looks at all of the pods running on that node for the day. Based on the resources the pod has reserved and the length of time it ran, the appropriate portion of the node's cost is assigned to that pod. This calculated cost is enriched with all of the pod's tags.

> Once all pods have been assigned a cost based on their resource reservations, some node cost is leftover. This is the cost of unreserved resources, which we call **Cluster Idle cost**. This cost is assigned the `is_cluster_idle` tag, and it represents the cost of resources that can be scheduled but are not reserved for any pods. For more information, see the [Understanding cluster idle cost](#understanding-cluster-idle-cost) section.

Reviewer: ```suggestion
Once all pods have been assigned a cost based on their resource reservations, some node cost is left over. This is the cost of unreserved resources, which is called **Cluster Idle cost**. This cost is assigned the `is_cluster_idle` tag, and it represents the cost of resources that can be scheduled but are not reserved for any pods. For more information, see the [Understanding cluster idle cost](#understanding-cluster-idle-cost) section.
```

### ECS on EC2

For ECS allocation, Datadog determines which tasks ran on each EC2 instance used for ECS.

Based on the CPU or memory usage of each task (as reported in the CUR), Datadog assigns the appropriate portion of the instance's compute cost to that task. The calculated cost is enriched with all of the task's tags and all of the container tags (except container names) for containers running in the task.

> Once all tasks have been assigned a cost based on their resource reservations, some instance cost is leftover. This is the cost of unreserved resources, which we call **Cluster Idle** cost. This cost is assigned the `is_cluster_idle` tag, and it represents the cost of resources not reserved by any ECS tasks. For more information, see the [Understanding cluster idle cost](#understanding-cluster-idle-cost) section.

Reviewer: ```suggestion
Once all tasks have been assigned a cost based on their resource reservations, some instance cost is left over. This is the cost of unreserved resources, which is called **Cluster Idle** cost. This cost is assigned the `is_cluster_idle` tag, and it represents the cost of resources not reserved by any ECS tasks. For more information, see the [Understanding cluster idle cost](#understanding-cluster-idle-cost) section.
```

### ECS on Fargate

> ECS tasks that run on Fargate are already fully allocated in the CUR. CCM enriches that data by adding out of the box tags and container tags to the AWS Fargate cost.

Reviewer: ```suggestion
ECS tasks that run on Fargate are already fully allocated in the CUR. CCM enriches that data by adding out-of-the-box tags and container tags to the AWS Fargate cost.
```

**Note**: For costs that are not related to containers on EC2, all three metrics are equal:
  - `aws.cost.amortized` 
  - `aws.cost.amortized.cpu.allocated`
  - `aws.cost.amortized.mem.allocated`

## Understanding cluster idle cost

Cluster idle cost is the cost of CPU or memory resources that are not reserved by any workloads.

You can visualize this cost at the Kubernetes node, ECS host, or cluster level, using the `is_cluster_idle` tag. All costs allocated to scheduled pods or tasks have `is_cluster_idle: N/A`, while all compute costs not assigned to any pods or tasks have the tag `is_cluster_idle: true`.

Here is a sample query that displays all Kubernetes cluster cost, broken down by cluster name and cluster idle cost:
   
`sum:aws.cost.amortized.cpu.allocated{is_kubernetes:true} by {kube_cluster_name, is_cluster_idle}`

  {{< img src="/cloud_cost/container_cost_allocation/cost_allocation_editor_config.png" alt="Example configuration for mixed resources with break down by cluster name and cluster idle cost tags" style="width:100%;" >}}

## Tags

Datadog applies additional tags to cost metrics where container costs are allocated.

### Kubernetes

> In addition to Kubernetes pod and Kubernetes node tags, the following out of the box tags are applied to cost metrics:

Reviewer: ```suggestion
In addition to Kubernetes pod and Kubernetes node tags, the following out-of-the-box tags are applied to cost metrics:
```

> | Out of the box tag  |  Description |

Reviewer: ```suggestion
| Out-of-the-box tag  |  Description |
```

| ---                 | ------------ | 

> | `kube_cluster_name` | The name of the kubernetes cluster. |

Reviewer: ```suggestion
| `kube_cluster_name` | The name of the Kubernetes cluster. |
```

| `is_kubernetes`     | All EC2 compute costs associated with running Kubernetes nodes. |
| `is_cluster_idle`   | The cost of unreserved CPU or memory on Kubernetes nodes. |

### ECS

> In addition to ECS task tags, the following out of the box tags are applied to cost metrics. **Note**: most tags from ECS container (excluding `container_name`) are applied.

Reviewer 3: Should `**Note**: most tags from ECS container (excluding `container_name`) are applied.` be `**Note**: most tags from ECS containers (excluding `container_name`) are applied.` ?

Author: ```suggestion
In addition to ECS task tags, the following out of the box tags are applied to cost metrics. **Note**: most tags from ECS containers (excluding `container_name`) are applied.
```

> In addition to ECS task tags, the following out of the box tags are applied to cost metrics. **Note**: most tags from ECS containers (excluding `container_name`) are applied.

Reviewer: ```suggestion
In addition to ECS task tags, the following out-of-the-box tags are applied to cost metrics. **Note**: most tags from ECS containers (excluding `container_name`) are applied.
```

> | Out of the box tag      |  Description |

Reviewer: ```suggestion
| Out-of-the-box tag      |  Description |
```
